orchagent/muxorch.cpp

```diff
@@ -480,70 +447,240 @@ bool MuxCable::isIpInSubnet(IpAddress ip)
         {
-            alias_ = alias;
+            return;
+        }
+
+        switch (state)
+        {
+        case MuxState::MUX_STATE_INIT:
+            neighbors_[nh.ip_address] = SAI_NULL_OBJECT_ID;
+            break;
+        case MuxState::MUX_STATE_ACTIVE:
+            neighbors_[nh.ip_address] = gNeighOrch->getLocalNextHopId(nh);
+            enable(false);
+            break;
+        case MuxState::MUX_STATE_STANDBY:
+            neighbors_[nh.ip_address] = tunnelId;
+            disable(tunnelId);
+            break;
+        default:
+            SWSS_LOG_NOTICE("State '%s' not handled for nbr %s update",
+                             muxStateValToString.at(state).c_str(), nh.ip_address.to_string().c_str());
+            break;
         }
     }
     else
     {
-        neighbors_.remove(ip);
+        MuxCableOrch* mux_cb_orch = gDirectory.get<MuxCableOrch*>();
+        IpPrefix pfx = nh.ip_address.to_string();
+
+        /* if current state is standby, remove the tunnel route */
+        if (state == MuxState::MUX_STATE_STANDBY)
+        {
+            remove_route(pfx);
+            mux_cb_orch->removeTunnelRoute(nh);
+        }
+        neighbors_.erase(nh.ip_address);
     }
 }
 
-bool MuxNbrHandler::enable()
+bool MuxNbrHandler::enable(bool update_rt)
 {
     NeighborEntry neigh;
+    MuxCableOrch* mux_cb_orch = gDirectory.get<MuxCableOrch*>();
 
-    auto it = neighbors_.getIpAddresses().begin();
-    while (it != neighbors_.getIpAddresses().end())
+    auto l_fn = [&] (NextHopKey nh_key, uint32_t val) {
+        SWSS_LOG_INFO("Inc ref for %s by %d", nh_key.ip_address.to_string().c_str(), val);
+        for (uint32_t cnt = 0; cnt < val; ++cnt)
+        {
+            gNeighOrch->increaseNextHopRefCount(nh_key);
+        }
+    };
+
+    auto it = neighbors_.begin();
+    while (it != neighbors_.end())
     {
-        neigh = NeighborEntry(*it, alias_);
+        SWSS_LOG_INFO("Enabling neigh %s on %s", it->first.to_string().c_str(), alias_.c_str());
+
+        neigh = NeighborEntry(it->first, alias_);
         if (!gNeighOrch->enableNeighbor(neigh))
         {
             return false;
         }
+
+        /* Update NH to point to learned neighbor */
+        it->second = gNeighOrch->getLocalNextHopId(neigh);
+
+        /* Reprogram route */
+        NextHopKey nh_key = NextHopKey(it->first, alias_);
+        uint32_t num_routes = 0;
+        if (!gRouteOrch->updateNextHopRoutes(nh_key, num_routes))
+        {
+            return false;
+        }
+
+        /* Increment ref count for new NHs */
+        l_fn (nh_key, num_routes);
+
+        /* Invalidate current nexthop group and update with new NH */
+        if (!gRouteOrch->invalidnexthopinNextHopGroup(nh_key, num_routes))
+        {
+            SWSS_LOG_ERROR("Removing existing NH failed for %s", nh_key.ip_address.to_string().c_str());
+            return false;
+        }
+
+        if (!gRouteOrch->validnexthopinNextHopGroup(nh_key, num_routes))
```

Comment:
both line 574 and 580 has num_routes, this cause readability of the code.

i think it is more readable to get del_route_num in line 574 and get add_route_num in line 580, and you can get their substraction add_route_num - del_route_num as refcnt change.

Reply:
In enable flow, it is moving from Tunnel NH -> Direct NH (Neighbor). In this case, the ref count must be incremented only for the Direct NH. Similarly for disable flow, ref count is decremented for Direct NH. However, there is no ref_cnt management for tunnel NH as it is one time created.